Adds validity checks to BLS public keys

brunoffranca opened the pull request:
## What ❔

Added checks for the validity of our BLS public keys as recommended by this [spec](https://www.ietf.org/archive/id/draft-irtf-cfrg-bls-signature-05.html). Namely we check that the public key is not zero and that it is in the correct subgroup.

We do it redundantly in two places (when we decode a public key and when we verify a signature).

## Why ❔

Invalid public keys are a security risk.

pompon0 commented:
Please add a test for the constructor.

dnkolegov reviewed (approved):
Great work @brunoffranca. Thanks.

brunoffranca commented:
@pompon0 I don't see a way of creating a point outside the correct subgroup. If I think of anything meanwhile, I'll add it.

pompon0 commented on node/libs/crypto/src/bn254/mod.rs:
```diff
@@ -132,6 +139,16 @@ impl Signature {
 
+        // Verify public key. Since we already check the validity of a
+        // public key when constructing it, this should never fail (in theory).
+        if pk.0.is_zero() {
```
nit `assert!(!pk.0.is_zero());`, analogically the other check

moshababo replied:
Why not verify prior to `hash_to_g1`?

moshababo replied:
Why not verify prior to `hash_to_g1()`?

pompon0 commented on node/libs/crypto/src/bn254/tests.rs:
```diff
@@ -89,6 +89,18 @@ fn aggregate_signature_failure_smoke() {
+    let pk_decoded = PublicKey::decode(&bytes);
+    let err = pk_decoded.unwrap_err();
+    assert_eq!(format!("{}", err.root_cause()), "Public key can't be zero");
+
```
don't compare human readable error messages. Just use `assert!(pk_decoded.is_err())`. Alternatively implement a decode() method which returns a strongly typed error that you may match against and make ByteFmt::decode just call your custom decode.

pompon0 commented on node/libs/crypto/src/bn254/tests.rs:
```diff
@@ -89,6 +89,18 @@ fn aggregate_signature_failure_smoke() {
+    let err = pk_decoded.unwrap_err();
+    assert_eq!(format!("{}", err.root_cause()), "Public key can't be zero");
+
+    // TODO: Create a G2 point in the wrong subgroup and test against it.
```
I suppose you can sample at random public key material, until you find a point in the wrong subgroup and then pass it to decode() to check that it also returns an error.

pompon0 replied:
or sample and hardcode into the test.

brunoffranca replied:
Nope. Random G2 generation always [creates points in the correct subgroup](https://github.com/matter-labs/pairing/blob/master/src/bn256/ec.rs#L1149).
There's a [test](https://github.com/matter-labs/pairing/blob/master/src/bn256/ec.rs#L1522) that I can port. I'll probably do it, just not today.

pompon0 replied:
not random g2, I mean random slice of bytes

moshababo replied:
@brunoffranca reading the code you referenced, it always creates a point from a random `x` because a loop is used. As mentioned in https://github.com/matter-labs/era-consensus/pull/80, it's about 50% chance.

brunoffranca commented:
So, there was a fatal flaw in the subgroup check. To check that a given point is in the correct subgroup, we don't multiply by the cofactor, we multiply by the order. I fixed that and found a way of producing testing it appropriately.
Another bug is that we were not doing the validity checks for aggregate signature verification. Also fixed it.
Finally, refactored it a bit to make it more readable.

pompon0 commented on node/libs/crypto/src/bn254/mod.rs:
```diff
@@ -96,10 +114,14 @@ impl Hash for PublicKey {
+        let pk = PublicKey(p.into());
+
+        if !pk.is_valid() {
+            anyhow::bail!("Public key is not valid")
```
anyhow::ensure!

pompon0 commented on node/libs/crypto/src/bn254/mod.rs:
```diff
@@ -81,6 +81,24 @@ impl PartialEq for SecretKey {
+impl PublicKey {
+    /// Checks if the public key is not the identity element and is in the correct subgroup. Verifying signatures
+    /// against public keys that are not valid is insecure.
+    pub fn is_valid(&self) -> bool {
```
it can be private, no?

pompon0 replied:
nit: more canonical would be to make it `fn verify(&self) -> Result<()>`, rather than return a bool.

pompon0 replied:
I think you should also add a check to SecretKey::decode(), because SecretKey::public() also constructs a PublicKey

brunoffranca replied:
It can be private, but I think it's better to keep it public. It's a useful method.

pompon0 replied:
I mean, all PublicKey objects ever constructed are expected to be valid, so having verify() in the public API will be quite misleading...

pompon0 commented on node/libs/crypto/src/bn254/mod.rs:
```diff
@@ -81,6 +81,24 @@ impl PartialEq for SecretKey {
+
 impl Default for PublicKey {
     fn default() -> Self {
         PublicKey(G2::zero())
```
It is problematic that we allow to construct an invalid public key. We should either remove the default constructor, or keep the is_valid/verify public and revert my suggestion of using `assert!` instead of returning an error.

pompon0 replied:
Ok, I've tried to remove the default constructor and it compiles just fine with minimal changes to AggregateSignature::verify.
